### PR TITLE
Feature/http client socket

### PR DESCRIPTION
**Pull request by moises-silva**
## Description
Refer to #1432 for details

## Related Issue
This implements the feature described in issue #1432 and supersedes #1352 

If merged, this would provide a suitable work-around for #1235 and also implement feature request #1304 

## Motivation and Context
Refer to #1432

## How Has This Been Tested?
Testing? what's that?

ok fine, I used curl to test all endpoints and I used netcat to test the raw socket still works.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

At this point I suppose I need feedback before moving on with unit testing/documentation etc

### CODE REVIEW

**Comment by moises-silva** [comment]
I'll have a look at the build failures, probably some tests broke due to moving process_check_result() into a utils file and no longer implicitly calling respond() inside (as that's specific to the socket client and process_check_result is used by both the http and raw socket client).

File: lib/sensu/client/http_socket.rb [hunk 1]
```diff
@@ -0,0 +1,208 @@
+require "base64"
+require "em-http-server"
+require "sensu/json"
+require "sensu/utilities"
+require "sensu/api/utilities/transport_info"
+require "sensu/client/utils"
+
+module Sensu
+  module Client
+    # EventMachine connection handler for the Sensu HTTP client"s socket.
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: ' vs "

File: lib/sensu/client/http_socket.rb [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    #
+    # The Sensu client listens on localhost, port 3031 (by default), for
+    # TCP HTTP connections. This allows software running on the host to
+    # push check results (that may contain metrics) into Sensu, without
+    # needing to know anything about Sensu"s internal implementation.
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: ' vs "

File: lib/sensu/client/http_socket.rb [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    # GET /info
+    #  This endpoint returns 200 OK with some basic sensu info
+    #
+    # POST /results
+    #  This endpoint expects application/json body with a check result
+    #
+    # GET /settings
+    #  This endpoint responds with 200 OK and the sensu configuration
+    #
+    # GET /brew
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: Haha, "this better return a 418 ... yup"

File: lib/sensu/client/http_socket.rb [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    # needing to know anything about Sensu"s internal implementation.
+    #
+    # All requests and responses expect a json-encoded body (if a body
+    # is expected at all).
+    #
+    # This socket requires receiving a proper HTTP request to any of
+    # the following endpoints:
+    #
+    # GET /info
+    #  This endpoint returns 200 OK with some basic sensu info
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: Sensu should probably be capitalized, not required, but a nicety.

File: lib/sensu/client/http_socket.rb [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            :transport => info
+          }
+          response.content = Sensu::JSON::dump(rdata)
+          response.send_response
+        end
+      end
+
+      def process_request_results(response)
+        @logger.debug("Processing results request")
+        if @http[:content_type] and @http[:content_type] == 'application/json' and @http_content
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: "'application/json" (double quotes) for extra style points :+1:

File: lib/sensu/client/http_socket.rb [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            check = Sensu::JSON::load(@http_content)
+            process_check_result(check)
+            rdata = {
+              :response => "ok"
+            }
+            response.content = Sensu::JSON::dump(rdata)
+            response.send_response
+          rescue Sensu::JSON::ParseError, ArgumentError
+            rdata = {
+              :response => "Failed to parse json body"
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: JSON

File: lib/sensu/client/http_socket.rb [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        end
+      end
+
+      def process_request_results(response)
+        @logger.debug("Processing results request")
+        if @http[:content_type] and @http[:content_type] == 'application/json' and @http_content
+          begin
+            check = Sensu::JSON::load(@http_content)
+            process_check_result(check)
+            rdata = {
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: Don't want to use `response_data` instead?

File: lib/sensu/client/http_socket.rb [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            :response => "Invalid content type"
+          }
+          response.content = Sensu::JSON::dump(rdata)
+          response.send_response
+        end
+      end
+
+      def process_request_settings(response)
+        if authorized?
+          @logger.debug("Processing settings request")
```
Comment (portertech) [inline lib/sensu/client/http_socket.rb]: The log events lack some context and should be all lowercase to match other events, e.g. `@logger.debug("processing http socket request", :endpoint => "/settings")`

**Comment by portertech** [comment]
@moises-silva a few minor changes to be made and a spec adjustment and this will be ready to be merged :+1: I'll assign this to the 0.27 milestone, will need to reopen/target the pr against the release/0.27 when it's ready :+1:

**Comment by portertech** [comment]
Closes https://github.com/sensu/sensu/issues/1432

**Comment by portertech** [comment]
Closes https://github.com/sensu/sensu/pull/1352

**Comment by portertech** [comment]
Closes https://github.com/sensu/sensu/issues/1235

File: lib/sensu/client/http_socket.rb (hunk 1 above)
Comment (moises-silva) [inline lib/sensu/client/http_socket.rb]: ha! good catch. Note that was copied from the existing socket.rb comments (I copied the whole block comment and adjusted for the HTTP endpoint, so socket.rb is still using the wrong quotes). I'll fix it here, but won't fix it in socket.rb to not muddle the waters. You have commit so I guess it's easy enough for you to fix it there if you want (or let me know if you want socket.rb fixed in this PR even though is unrelated).

File: lib/sensu/client/http_socket.rb (hunk 4 above)
Comment (moises-silva) [inline lib/sensu/client/http_socket.rb]: agreed, I'll fix all instances of this in comments and logs

File: lib/sensu/client/http_socket.rb (hunk 7 above)
Comment (moises-silva) [inline lib/sensu/client/http_socket.rb]: I can do much better, wait for my update, it'll remove much duplication and looks cleaner IMO

File: lib/sensu/client/http_socket.rb (hunk 8 above)
Comment (moises-silva) [inline lib/sensu/client/http_socket.rb]: yeah I added most of those when still confirming my routes were working, I think now it's just better to stick to the one log statement when processing starts, we're already logging the endpoint there

**Comment by moises-silva** [comment]
@portertech ok I think things from my end are ok, just lacking any sort of tests for this. I may be able to find some time to work on them over the weekend.

**Comment by portertech** [comment]
Will have another good look at this PR and the associated documentation this evening :+1:

**Comment by portertech** [comment]
Going to merge this, even tho this PR is missing additional tests to cover the new functionality/code. I will push a PR later today with such tests.

**Comment by portertech** [comment]
@moises-silva great work :+1:  Thank you for all of your effort to get this into Sensu core.

**Comment by moises-silva** [comment]
yay! sweet

**Pull request merged**